gateway-server/src/main/java/org/apache/knox/gateway/services/token/impl/AliasBasedTokenStateService.java

```diff
@@ -122,6 +129,44 @@ public void start() throws ServiceLifecycleException {
                                                     statePersistenceInterval,
                                                     TimeUnit.SECONDS);
     }
+
+    // Loading ALL entries from __gateway-credentials.jceks could be VERY time-consuming (it took a bit more than 19 minutes to load 12k aliases
+    // during my tests).
+    // Therefore, it's safer to do it in a background thread than just make the service start hang until it's finished
+    final ExecutorService gatewayCredentialsLoader = Executors.newSingleThreadExecutor(new BasicThreadFactory.Builder().namingPattern("GatewayCredentialsLoader").build());
+    gatewayCredentialsLoader.execute(this::loadGatewayCredentialsOnStartup);
+  }
+
+  private void loadGatewayCredentialsOnStartup() {
+    try {
+      log.loadingGatewayCredentialsOnStartup();
+      final long start = System.currentTimeMillis();
+      final Map<String, char[]> passwordAliasMap = aliasService.getPasswordAliasMapForGateway();
+      String alias, tokenId;
+      long expiration, maxLifeTime;
+      int count = 0;
+      for (Map.Entry<String, char[]> passwordAliasMapEntry : passwordAliasMap.entrySet()) {
+        alias = passwordAliasMapEntry.getKey();
+        if (alias.endsWith(TOKEN_MAX_LIFETIME_POSTFIX)) {
```

Comment:
Some comments here might be helpful for future viewers. I understand what is being done here, but others may not.

Reply:
Done.